README.md
# create-react-native-module

Tool to create a React Native library module or native view component, with a single command.

[![GitHub license](https://img.shields.io/github/license/brodybits/create-react-native-module.svg?color=blue&style=for-the-badge)](./LICENSE)
[![npm](https://img.shields.io/npm/v/create-react-native-module.svg?color=green&style=for-the-badge)](https://www.npmjs.com/package/create-react-native-module)
[![Mutation testing badge](https://img.shields.io/endpoint?style=for-the-badge&url=https%3A%2F%2Fbadge-api.stryker-mutator.io%2Fgithub.com%2Fbrodybits%2Fcreate-react-native-module%2Fmaster)](https://dashboard.stryker-mutator.io/reports/github.com/brodybits/create-react-native-module/master)
[![npm downloads](https://img.shields.io/npm/dw/create-react-native-module.svg?label=npm%20downloads&style=for-the-badge)](https://npmcharts.com/compare/create-react-native-module?minimal=true)
[![total npm downloads](https://img.shields.io/npm/dt/create-react-native-module.svg?label=total%20npm%20downloads&style=for-the-badge)](https://npmcharts.com/compare/create-react-native-module?minimal=true)
[![GitHub watchers](https://img.shields.io/github/watchers/brodybits/create-react-native-module.svg?style=for-the-badge)](https://github.com/brodybits/create-react-native-module/watchers)
[![GitHub stars](https://img.shields.io/github/stars/brodybits/create-react-native-module.svg?label=GitHub%20stars&style=for-the-badge)](https://github.com/brodybits/create-react-native-module/stargazers)
[![GitHub forks](https://img.shields.io/github/forks/brodybits/create-react-native-module.svg?style=for-the-badge)](https://github.com/brodybits/create-react-native-module/network/members)
[![open bugs](https://img.shields.io/github/issues-raw/brodybits/create-react-native-module/bug.svg?color=d73a4a&label=open%20bugs&style=for-the-badge)](https://github.com/brodybits/create-react-native-module/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+label%3Abug)
[![total open issues](https://img.shields.io/github/issues-raw/brodybits/create-react-native-module.svg?label=total%20open%20issues&style=for-the-badge)](https://github.com/brodybits/create-react-native-module/issues)
[![GitHub pull requests](https://img.shields.io/github/issues-pr-raw/brodybits/create-react-native-module.svg?style=for-the-badge)](https://github.com/brodybits/create-react-native-module/pulls)

See below for command-line usage, example with no view, and example with an extremely simple native view.

This tool based on [`react-native-create-library`](https://www.npmjs.com/package/react-native-create-library), with working example callbacks, optional native view, and more updates added by [@brodybits (Christoper J. Brody aka Chris Brody)](https://github.com/brodybits) and other [contributors](https://github.com/brodybits/create-react-native-module/graphs/contributors).

### Support options

- community support via [issues](https://github.com/brodybits/create-react-native-module/issues)
- commercial support is available, see <http://xpbrew.consulting>

<!-- FUTURE TODO: QUICK USAGE SECTION HERE -->

### General status

- Minimum React Native version: `0.60` (outdated), `0.63` (recommended) - see issue [`#423`](https://github.com/brodybits/create-react-native-module/issues/423)
- It is recommended to disable or remove Flipper on iOS in case of debug build issues - see issue [`#422`](https://github.com/brodybits/create-react-native-module/issues/422)
- generated example app with symlink by default, has known issue with adding dependencies to the library root - see issue [`#308`](https://github.com/brodybits/create-react-native-module/issues/308)
- Platform fork support
  - tvOS platform fork
    - requires use of `--tvos-enabled` option as documented below
    - requires the [`react-native-tvos`](https://www.npmjs.com/package/react-native-tvos) fork, with minimum version of `0.60` ref:
      - [`react-native-tvos/react-native-tvos#11`](https://github.com/react-native-tvos/react-native-tvos/issues/11)
      - issue [`#95`](https://github.com/brodybits/create-react-native-module/issues/95)
    - It is recommended to disable or remove Flipper in case of debug build issues - see issue [`#422`](https://github.com/brodybits/create-react-native-module/issues/422)
- Out-of-tree platforms
  - Windows - no longer supported for reasons discussed in issues [#23](https://github.com/brodybits/create-react-native-module/issues/23) and [#43](https://github.com/brodybits/create-react-native-module/issues/43) (existing Windows C# template is kept in `unsupported-platforms` for now (at least) and further discussion would be welcome in a new issue on GitHub)
  - for future consideration: macOS (see [issue #94](https://github.com/brodybits/create-react-native-module/issues/94))

### Why might you need this?

If you are looking to create a native module for React Native, you need some native code for each platform you want to support and then some JavaScript code to bind it all together. Setting this up by yourself can be time-consuming.

This is where this tool comes in. It creates a boilerplate with all current best practices in mind.
Why not use `react-native new-library`? Unfortunately that command doesn't create an up-to-date library, requires an already initialized React Native project and only sets up the iOS side of things.

### Alternatives

- [`otobank/create-react-native-module`](https://github.com/otobank/create-react-native-module) - fork with TypeScript, Swift, and Kotlin support, with some other updates and some less useful features removed; see also issue [`#425`](https://github.com/brodybits/create-react-native-module/issues/425)
- [`brodybits/react-native-module-init`](https://github.com/brodybits/react-native-module-init) - new interactive CLI that uses the templates from this utiity
- [`react-native-community/bob`](https://github.com/react-native-community/bob) - opinionated, interactive library CLI that is designed to support both native libraries and libraries with web support

__Outdated alternatives:__ see [acknowledgements](#acknowledgements) below

## Installation

Package required to be installed globally if the recommended example app is generated:

- [`yarn`](https://www.npmjs.com/package/yarn)

```console
$ npm install -g yarn
```

To install this package:

```console
$ npm install -g create-react-native-module
```

## Command-line usage

Navigate into an empty directory to execute the command.

```console
$ create-react-native-module MyFancyLibrary
```

This will create the folder `MyFancyLibrary` in which the library will be created in.

Now install dependencies by running this command in the newly created library.

```console
$ npm install
```

```console
Usage: create-react-native-module [options] <name>

Options:

  -V, --version                             output the version number
  --package-name <packageName>              The full package name to be used in package.json. Default: react-native-(name in param-case)
  --is-view                                 Generate the package as a very simple native view component. Status: EXPERIMENTAL, with limited testing.
  --object-class-name                       The name of the object class to be exported by both JavaScript and native code. Default: (name in PascalCase)
  --native-package-id <nativePackageId>     [Android] The native Java package identifier used for Android (Default: `com.reactlibrary`)
  --platforms <platforms>                   Platforms the library module will be created for - comma separated (Default: `ios,android`)
  --tvos-enabled                            Generate the module with tvOS build enabled (requires react-native-tvos fork, with minimum version of 0.60, and iOS platform to be enabled)
  --github-account <githubAccount>          The github account where the library module is hosted (Default: `github_account`)
  --author-name <authorName>                The author's name (Default: `Your Name`)
  --author-email <authorEmail>              The author's email (Default: `yourname@email.com`)
  --license <license>                       The license type (Default: `MIT`)
  --use-apple-networking                    [iOS] EXPERIMENTAL FEATURE NOT SUPPORTED: Use `AFNetworking` dependency as a sample in the podspec & use it from the iOS code - see issue #426
  --generate-example                        Generate an example project and add the library module to it with symlink by defult, with overwrite of example metro.config.js to add workaround for Metro symlink issue - requires Yarn to be installed globally
  --example-file-linkage                    DEPRECATED: do `yarn add file:../` instead of `yarn add link:../` in a generated example project, and add a postinstall workaround script, with no overwrite of example metro.config.js
  --example-name <exampleName>              Name for the example project (default: `example`)
  --example-react-native-template <...>     The React Native template used for the generated example project, for example: react-native-tvos or react-native-tvos@0.62.2-1 (requires --tvos-enabled option); react-native@0.62 (default: `react-native@latest`)
  --write-example-podfile                   [iOS] EXPERIMENTAL FEATURE NOT SUPPORTED: write (or overwrite) example ios/Podfile
  -h, --help                                output usage information
```

## Programmatic usage

```javascript
const createLibraryModule = require('create-react-native-module');

createLibraryModule({
  name: 'MyFancyLibraryModule'
}).then(() => {
  console.log('Oh yay! My library module has been created!');
})
```

#### Options

```javascript
{
  name: String, /* The name of the library (mandatory) */
  packageName: String, /* The full package name to be used in package.json. Default: react-native-(name in param-case) */
  isView: Boolean, /* Generate the package as a very simple native view component. Status: EXPERIMENTAL, with limited testing. (Default: false) */
  objectClassName: String, /* The name of the object class to be exported by both JavaScript and native code. Default: (name in PascalCase) */
  platforms: Array | String, /* Platforms the library will be created for. (Default: ['android', 'ios']) */
  nativePackageId: String, /* [Android] The native Java package identifier used for Android (Default: `com.reactlibrary`) */
  tvosEnabled: Boolean, /* Generate the module with tvOS build enabled (requires react-native-tvos fork, with minimum version of 0.60, and iOS platform to be enabled) */
  githubAccount: String, /* The github account where the library is hosted (Default: `github_account`) */
  authorName: String, /* The author's name (Default: `Your Name`) */
  authorEmail: String, /* The author's email (Default: `yourname@email.com`) */
  license: String, /* The license type of this library (Default: `MIT`) */
  useAppleNetworking: Boolean, /* [iOS] EXPERIMENTAL FEATURE NOT SUPPORTED: Use `AFNetworking` dependency as a sample in the podspec & use it from the iOS code - see issue #426 (Default: false) */
  generateExample: Boolean, /* Generate an example project and add the library module to it with symlink by defult, with overwrite of example metro.config.js to add workaround for Metro symlink issue - requires Yarn to be installed globally (Default: false) */
  exampleFileLinkage: Boolean, /* DEPRECATED: do `yarn add file:../` instead of `yarn add link:../` in a generated example project, and add a postinstall workaround script, with no overwrite of example metro.config.js (Default: false) */
  exampleName: String, /* Name for the example project (Default: `example`) */
  exampleReactNativeTemplate: String, /* The React Native template used for the generated example project, for example: react-native-tvos or react-native-tvos@0.62.2-1 (requires --tvos-enabled option); react-native@0.62 (Default: `react-native@latest`) */
  writeExamplePodfile: Boolean, /* [iOS] EXPERIMENTAL FEATURE NOT SUPPORTED: write (or overwrite) example ios/Podfile (Default: false) */
}
```

## Examples

### Example module with no view

__Create the module with no view:__

```console
create-react-native-module --package-identifier io.mylibrary --generate-example AliceHelper
```

The module would be generated in the `react-native-alice-helper` subdirectory, and the example test app would be in `react-native-alice-helper/example`.

Then go into the example app subdirectory:

```console
cd react-native-alice-helper/example
```

#### Running the example app

**Recommended:** Follow the instructions shown in the end of the console log output, which are more likely to be up-to-date.

__Extra notes:__

_Within the example test app subdirectory:_

It is *recommended* to start the Metro Bundler manually (within `react-native-alice-helper/example`), which would run in the foreground:

```console
yarn start
```

Otherwise, React Native will open its own window to run the Metro Bundler.

To run on Android, do the following command (within `react-native-alice-helper/example`):

```console
yarn android
```

for React Native `0.60`: `npx react-native run-android`

This assumes that the `ANDROID_HOME` environmental variable is set properly. Here is a sample command that does not make such an assumption on a mac:

```console
ANDROID_HOME=~/Library/Android/sdk npx react-native run-android
```

For iOS:

Extra installation step needed _in case of clean checkout only_:

```console
cd ios && pod install && cd ..
```

Then to run on iOS:

```console
yarn ios
```

for React Native `0.60`: `npx react-native run-ios`

or do the following command to open the iOS project in Xcode:

```console
open ios/example.xcodeproj
```

#### Expected result

The example app shows the following indications:

- STATUS: native callback received
- NATIVE CALLBACK MESSAGE with the number argument and string argument values that are received by the native module

### Example view module

EXPERIMENTAL feature with limited testing

__Create the module with an extremely simple view:__

```console
create-react-native-module --package-identifier io.mylibrary --is-view --generate-example CarolWidget
```

The module would be generated in the `react-native-carol-widget` subdirectory, and the example test app would be in `react-native-carol-widget/example`.

Then go into the example app subdirectory:

```console
cd react-native-carol-widget/example
```

#### Running the view example app

**Recommended:** Follow the instructions shown in the end of the console log output, which are more likely to be up-to-date.

__Some extra notes:__

_Within the example test app subdirectory:_

It is *recommended* to start the Metro Bundler manually as described above (within `react-native-carol-widget/example`):

```console
yarn start
```

To run on Android: do `yarn android` or `npx react-native run-android` as described for the other example above.

To run on iOS (as described above):

- _in case of clean checkout **only**_: do `pod install` in `ios` subdirectory
- do `yarn ios`, `npx react-native run-ios`, or `open ios/example.xcodeproj`

__Expected result:__

- on Android: a check box that is checked (and cannot be changed)
- on iOS: a label with 5 red asterisks

## Acknowledgements

- [`react-native-create-library`](https://www.npmjs.com/package/react-native-create-library) - original basis of this project
- [`react-native-share`](https://www.npmjs.com/package/react-native-share) - was acknowledged as "a great source of inspiration" for [`react-native-create-library`](https://www.npmjs.com/package/react-native-create-library)

## License

[MIT](./LICENSE)


CHANGELOG.md
# Changelog

### 0.20.2

* fix: add `CHANGELOG.md` to files published by npm (and explicitly include `LICENSE`)
* fix: [nit] exclude `.eslint*.*` from npm published files (was part of `templates`)

### 0.20.1

* update package dependencies
  - update execa -> ^5.1.1
  - update fs-extra -> ^10.0.0
  - update react-native-init-func -> ^0.0.6

### 0.20.0

#### Features

* BREAKING: rename some options (#431 & #439)
  - rename --module-name option to --package-name (#431)
  - rename --package-identifier option to --native-package-id (#431)
  - rename --example-react-native-version to --example-react-native-template (#431)
  - rename --view to --is-view; document this feature as EXPERIMENTAL; and update doc & help text (#439)
* BREAKING: remove prefix options (#432)
  - remove --prefix option
  - remove --module-prefix option
* BREAKING: use react-native-init-func to generate example app (#421)
* support more iOS source language extensions in generated podspec (#410)

#### Fixes

* loosen template React peer dependency (for React 17) (#440)
* nudge indenting of options in logging (#436)
* do not add react-native package prefix if not needed (#435)
* update example build & run info in README.md & log output (#427)
* remove an internal Promise reject parameter not needed (#417)
* use info for some more log output (#418)
* add comments to generated output
  - add comment to main generated index.js (#413)
  - add comments to generated Android output (#137)
  - add comments to beginning of generated iOS output (#414)
* use outdent to fix indent of optional example Podfile (#412)
* resolve absolute paths in example metro watch folders (#405)
* add & update comments in build.gradle (#404)
* remove extra plugin items from build.gradle (#402)
* BREAKING: remove JCenter support - not needed and now deprecated (#401)
* remove extra import from generated code for Android (#368)
* BREAKING: update minimum Node.js version to 12 (#408)
* update package dependencies
  - update log-symbols -> ^4.1.0 (#438)
  - update update-notifier -> 5 (commit fb8542ee05295523b7f751487bf150efae1afc5b)
  - update execa -> 5 (commit a3a7ab927d338435ef281baba05ef69888fc5fde)
  - Update fs-extra -> ^9.1.0 (#398)
  - Update pascal-case -> ^3.1.2 (#395)
  - Update param-case -> ^3.0.4 (#394)
  - Update jsonfile -> ^6.1.0 (#391)
  - Update commander -> ^6.2.1 (#382)
* add & use console package (#375)
* minor template cleanup (#332)

#### Docs

* update general status for React Native 0.60/0.63/0.64 (commit 23bdce59b840687b08268a6daa2991fa04b4802c)
* console highlighting in README.md (#434)
* apply general documentation updates - February 2021 (#424)

### 0.19.0

* remove outdated info from README.md (#355)
* remove extra blank line from generated iOS module (#354)
* Update execa -> ^4.0.1 - dependency (#350)
* add swift search paths to iOS pbxproj (#349)
* add pbxproj build updates (#348)
* remove sources from generated iOS pbxproj (#346)
* fix minor indenting issue in ios pbxproj (#345)
* export objectClassName prop in windows template (#344)
* cleanup object class name references (#343)
* log generated example React Native version (#340)
* major: object class name updates (#338)
* fix iOS podspec for C sources (#334)
* major: remove generated .gitattributes artifact (#327)
* use package files entry instead of .npmignore (#326)
* update create options (#325)
* update badges in README.md (#322)
* update log of create parameters (#318)
* Update jsonfile -> ^6.0.1 (#317)
* update commander -> ^5.1.0
* update log-symbols -> 4
* update fs-extra -> 9
* fix(deps): update execa -> 4 (#315)
* major: update minimum Node.js version to 10.20 (#316)

### 0.18.0

* generate example with symlink by default, with multiple workarounds in example metro.config.js (#309)

### 0.17.1

* log & document issue with example lib dependencies (#310)

### 0.17.0

* bugfix: quick fixes to handle CLI args errors (#306)
* test cleanup: rename tests/with-mocks/cli/command/func directory
* chore: regenerate yarn.lock
* chore: jest@^25.2.7 update in devDependencies
* add & update iOS podspec license comments (#258)
* doc: minimum React Native version on 1 line (#302)
* doc: fix API & option descriptions (#301)
* chore: update change case dependencies (#298)
* chore: quick commander@5 update (#297)
* chore: add PR merge action (#296)
* chore: jest@25 update in devDependencies
* chore: update eslint items in devDependencies
* chore: @stryker-mutator/... 3.1.0 in devDependencies
* chore: add `brodybits-*` to test action branches
* add package engines & use please-upgrade-node (#295)
* cleanup: remove .travis.yml which is no longer needed
* testing: add Node.js test action to GitHub workflows (#294)
* testing: quick test workaround for Node.js 8 vs ...
* cleanup: remove library API support for namespace (#276)
* testing: updates to setup of Stryker Mutator (#272)
* remove Windows (C#) support (#264)


### 0.16.0

* use react-native@latest in example by default (#263)
* doc: update alternatives in README.md (#262)
* cleanup: refactoring: move `require('uuid')` out of templates (#261)
* doc: example app doc updates (quick updates) (#260)
* log more example app info in the end (#259)

### 0.15.0

* no log of generated example options unless enabled (#257)
* show hint to generate example, if needed in CLI (#256)
* chore: Bump jsonfile from 5.0.0 to 6.0.0 (#252)
* doc: fix default React Native version in lib API (#255)
* chore: Bump update-notifier from 3.0.1 to 4.1.0 (#250)

### 0.14.1

* bugfix: fix license in iOS podspec (#253)

### 0.14.0

* support React Native 0.61; drop React Native 0.59 (#244)
* chore: uuid 3.4.0 update in package dependencies (#243)
* chore: [dev] @stryker-mutator 2.5.0 updates (#242)
* chore: [dev] eslint-plugin-node@11 update (#241)
* chore: Bump eslint-plugin-import from 2.18.2 to 2.20.1 (#234)
* chore: Bump eslint from 6.6.0 to 6.8.0 (#215)
* chore: [Security - dev] Bump handlebars from 4.1.2 to 4.7.3 (#235)

lib/cli-command.js
const { info, error } = require('console');

const emoji = require('node-emoji');

const logSymbols = require('log-symbols');

const normalizedOptions = require('./normalized-options');

const createLibraryModule = require('./lib');

const postCreateInstructions = ({
  packageName,
  platforms,
  generateExample,
  exampleName
}) => `
====================================================
YOU'RE ALL SET!
` + (generateExample
    ? `
${emoji.get('bulb')} check out the example app in ${packageName}/${exampleName}
${emoji.get('bulb')} recommended: run Metro Bundler in a new shell
${logSymbols.info} (cd ${packageName}/${exampleName} && yarn start)
${emoji.get('bulb')} enter the following commands to run the example app:
${logSymbols.info} cd ${packageName}/${exampleName}
${platforms.split(',').map(platform =>
  `${logSymbols.info} yarn ${platform} # for React Native 0.60: npx react-native run-${platform}`
).join(`
`)}
${logSymbols.warning} IMPORTANT NOTICES
${logSymbols.warning} After clean checkout, these first steps are needed:
${logSymbols.info} run Yarn in ${packageName}/${exampleName}/ios
${logSymbols.info} (cd ${packageName}/${exampleName} && yarn)
${logSymbols.info} do \`pod install\` for iOS in ${packageName}/${exampleName}/ios
${logSymbols.info} cd ${packageName}/${exampleName}
${logSymbols.info} (cd ios && pod install)
${logSymbols.warning} KNOWN ISSUE with adding dependencies to the library root
${logSymbols.info} see https://github.com/brodybits/create-react-native-module/issues/308
`
    : `
${emoji.get('bulb')} next time consider using \`--generate-example\` to add a generated example!
`);

module.exports = {
  name: 'create-library',
  description: 'creates a React Native library module for one or more platforms',
  usage: '[options] <name>',
  action: (args, options) => {
    if (!args || args.length < 1) {
      throw new Error('missing lib module name');
    }

    if (args.length > 1) {
      throw new Error('too many arguments');
    }

    const name = args[0];

    const beforeCreation = Date.now();

    const preNormalizedOptions = Object.assign({}, { name }, options);

    // NOTE: There is a trick where the new normalizedOptions()
    // from normalized-options.js is applied by both command.js & lib.js.
    // This is to ensure that the CLI gets the correct module name for the
    // final log message, and that the exported programmatic
    // function can be completely tested from using the CLI.

    const createOptions = normalizedOptions(preNormalizedOptions);

    const rootModuleName = createOptions.packageName;

    return createLibraryModule(createOptions).then(() => {
      info(`
${emoji.get('books')}  Created library module ${rootModuleName} in \`./${rootModuleName}\`.
${emoji.get('clock9')}  It took ${Date.now() - beforeCreation}ms.
${postCreateInstructions(createOptions)}`);
    }).catch((err) => {
      error(`Error while creating library module ${rootModuleName}`);

      if (err.stack) {
        error(err.stack);
      }
    });
  },
  options: [{
    command: '--package-name [packageName]',
    description: 'The full package name to be used in package.json. Default: react-native-(name in param-case)',
  }, {
    command: '--is-view',
    description: 'Generate the package as a very simple native view component. Status: EXPERIMENTAL, with limited testing.',
  }, {
    command: '--object-class-name [objectClassName]',
    description: 'The name of the object class to be exported by both JavaScript and native code. Default: (name in PascalCase)',
  }, {
    command: '--native-package-id [nativePackageId]',
    description: '[Android] The native Java package identifier used for Android',
    default: 'com.reactlibrary',
  }, {
    command: '--platforms <platforms>',
    description: 'Platforms the library module will be created for - comma separated',
    default: 'ios,android',
  }, {
    command: '--tvos-enabled',
    description: 'Generate the module with tvOS build enabled (requires react-native-tvos fork, with minimum version of 0.60, and iOS platform to be enabled)',
  }, {
    command: '--github-account [githubAccount]',
    description: 'The github account where the library module is hosted',
    default: 'github_account',
  }, {
    command: '--author-name [authorName]',
    description: 'The author\'s name',
    default: 'Your Name',
  }, {
    command: '--author-email [authorEmail]',
    description: 'The author\'s email',
    default: 'yourname@email.com',
  }, {
    command: '--license [license]',
    description: 'The license type',
    default: 'MIT',
  }, {
    command: '--use-apple-networking',
    description: '[iOS] Use `AFNetworking` dependency as a sample in the podspec & use it from the iOS code',
  }, {
    command: '--generate-example',
    description: 'Generate an example project and add the library module to it with symlink by defult, with overwrite of example metro.config.js to add workaround for Metro symlink issue - requires Yarn to be installed globally',
  }, {
    command: '--example-file-linkage',
    description: "DEPRECATED: do `yarn add file:../` instead of `yarn add link:../` in a generated example project, and add a postinstall workaround script, with no overwrite of example metro.config.js",
  }, {
    command: '--example-name [exampleName]',
    description: 'Name for the example project',
    default: 'example',
  }, {
    command: '--example-react-native-template [exampleReactNativeTemplate]',
    description: 'The React Native template used for the generated example project, for example: react-native-tvos or react-native-tvos@0.62.2-1 (requires --tvos-enabled option); react-native@0.62',
    default: 'react-native@latest',
  }, {
    command: '--write-example-podfile',
    description: '[iOS] EXPERIMENTAL FEATURE NOT SUPPORTED: write (or overwrite) example ios/Podfile',
  }]
};
